Lib/xml/etree/ElementTree.py

```diff
@@ -1256,21 +1256,25 @@ def iterator(source):
                 it.root = root
         finally:
             if close_source:
-                source.close()
+                1#source.close()
 
     gen = iterator(source)
     class IterParseIterator(collections.abc.Iterator):
         __next__ = gen.__next__
+
         def close(self):
+            nonlocal close_source
             if close_source:
                 source.close()
+                close_source = False
             gen.close()
 
-        def __del__(self):
-            # TODO: Emit a ResourceWarning if it was not explicitly closed.
-            # (When the close() method will be supported in all maintained Python versions.)
+        def __del__(self, _warn=warnings.warn):
             if close_source:
-                source.close()
+                try:
+                    _warn(f"unclosed iterparse iterator {source.name!r}", ResourceWarning, stacklevel=2)
+                finally:
+                    source.close()
```

Comment:
Maybe this is better?
```suggestion
                name = getattr(source, 'name', None)
                if name:
                    _warn("unclosed iterparse iterator %r" % (name,), 
                          ResourceWarning, stacklevel=2)
                source.close()
```

Reply:
`source.name` always exist and not empty. This code is only executed when `iterparse()` opened a file by name.

BTW, if the `iterparse()` argument was file descriptor 0, your variant would not work.

Reply:
You're right - I missed that `close_source=True` guarantees `source.name` exists, and `if name:` breaks on fd 0. Thanks for catching that! 

The original code is correct. (Though % formatting might be slightly safer than f-string in `__del__`, but not a blocker.)